### DISCUSSION TITLE
Failed to record new usage data

### DISCUSSION THREAD

**Discussion by GreatHolyCow** [jertel/vuegraf#42] [General]
2021-05-08 20:18:45.970045 | ERROR | Failed to record new usage data: (<class 'TypeError'>, TypeError("argument of type 'NoneType' is not iterable"), <traceback object at 0xb4d378f0>)

This started happening 5-7-21 afternoon. I did grab the latest at the time of this post to see if that would fix it and the error still persists.

**Comment by jertel** [comment]
Mine is still humming along. Any chance your account could have been modified? Assuming that error message is from the latest version of Vuegraf, my best guess from that error message is vuegray.py line 195 is returning None instead of a list of channels. Check your Emporia app and make sure everything looks correct on there. If it does, the next step would be to try some of the diagnostic Python scripting that is in #41 and see if you can reproduce that error outside of Vuegraf.

**Reply by GreatHolyCow** [reply]
So I have a Vue Gen 2 and 4 plugs, the app seems fine... I stopped using one of the plugs so I removed it from my account and now the script seems to work...

**Reply by jertel** [reply]
When you updated to the latest vuegraf do you remember if Vuegraf continued on after logging that error?

**Reply by GreatHolyCow** [reply]
No it would keep logging that error every time it was attempting to input data...

**Reply by pc-99** [reply]
I had the same issue, removing my unplugged plug from my account fixed it

**Reply by jertel** [reply]
I've pushed a change that adds a stack trace to better help me understand what's going on with your environments. Also included some `If ... is not None` protections but it's unclear it those will help you.

**Reply by pc-99** [reply]
One of my plug devices went offline last night, here's the additional logging:
```
2021-05-20 13:00:04.424380 | ERROR | Failed to record new usage data: (<class 'TypeError'>, TypeError("argument of type 'NoneType' is not iterable",), <traceback object at 0x7fc3b285e788>)
Traceback (most recent call last):
  File "/home/<user>/vuegraf/src/vuegraf.py", line 196, in <module>
    channels = account['vue'].get_devices_usage(deviceGids, None, scale=Scale.DAY.value, unit=Unit.KWH.value)
  File "/home/<user>/.local/lib/python3.6/site-packages/pyemvue/pyemvue.py", line 88, in get_devices_usage
    if channel: channels.append(VueDeviceChannelUsage().from_json_dictionary(channel))
  File "/home/<user>/.local/lib/python3.6/site-packages/pyemvue/device.py", line 97, in from_json_dictionary
    if 'value' in js['usage']:
TypeError: argument of type 'NoneType' is not iterable
```

**Reply by jertel** [reply]
@magico13, any idea why this is happening? I'm looking at your device.py file, around line 97 and it looks like line 96 already has protections in place in case a `None` got into js['usage']:

![image](https://user-images.githubusercontent.com/16639611/118985026-17083900-b94c-11eb-9278-6816e5d7348e.png)

**Reply by magico13** [reply]
@jertel Line 96 got updated in 0.12.4 to add the extra checking for None, specifically the `and js['usage']` part, so if you're not using that version try that. I haven't figured out why some people are seeing some channels/devices randomly returning no data after its been working fine for some time.

**Reply by jertel** [reply]
Thanks. The latest Vuegraf release uses 0.12.4. @pc-99 are you using Python directly, or are you using the Docker image?

**Reply by pc-99** [reply]
Yes, I run python directly in a screen, I updated pyemvue to 0.12.4 and it looks like it working now when I unplug a plug.  This happened to me before(not updating pyemvue), I'll make a note to update both from now on.  Sorry about that and thanks for the help.
